plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt

```diff
@@ -79,7 +80,42 @@ abstract class LoginBrowser(
 
     protected val selectionSettings = mutableMapOf<String, BearerConnectionSelectionSettings>()
 
-    protected val onPendingToken: (InteractiveBearerTokenProvider) -> Unit = { provider ->
+    private var browserOpenTimer: Timer? = null
+
+    private fun startBrowserOpenTimer(credentialSourceId: CredentialSourceId) {
+        browserOpenTimer = Timer()
+        browserOpenTimer?.schedule(
+            object : TimerTask() {
+                override fun run() {
+                    AwsTelemetry.loginWithBrowser(
+                        project = null,
+                        credentialStartUrl = SONO_URL,
+                        result = Result.Failed,
+                        reason = "Browser authentication idle for more than 15min",
+                        credentialSourceId = credentialSourceId
+                    )
+                    AuthTelemetry.addConnection(
+                        result = Result.Failed,
+                        reason = "Browser authentication idle for more than 15min",
+                        credentialSourceId = credentialSourceId
+                    )
+                    stopBrowserOpenTimer()
+                }
+            },
+            900000, // 15min
+        )
+    }
+
+    protected fun stopBrowserOpenTimer() {
+        if (browserOpenTimer != null) {
+            browserOpenTimer?.cancel()
+            browserOpenTimer?.purge()
+            browserOpenTimer = null
+        }
+    }
+
+    protected fun getOnPendingToken(credentialSourceId: CredentialSourceId): (InteractiveBearerTokenProvider) -> Unit = { provider ->
+        startBrowserOpenTimer(credentialSourceId)
```

Comment:
can't you infer the credential source from the provider?

Reply:
We could infer it from the [display name](https://github.com/aws/aws-toolkit-jetbrains/blob/95e0fc0b98d9918dc842b332bc4acd2ed04dca5b/plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/credentials/sso/bearer/BearerTokenProvider.kt#L99), which essentially involves string matching on start url (displayName is constructed [here](https://github.com/aws/aws-toolkit-jetbrains/blob/95e0fc0b98d9918dc842b332bc4acd2ed04dca5b/plugins/core/core/src/software/aws/toolkits/core/credentials/ToolkitCredentialsProvider.kt#L134)). But I feel explicitly identifying the credential source is more robust and doesn't add too much complexity

Reply:
`InteractiveBearerTokenProvider` has a startUrl which would also be trivial